Add MAGMA implementations of Torch LAPACK functions

colesbury opened the pull request:
This adds MAGMA binding for a number of LAPACK functions. If MAGMA is not installed, the functions simply raise an error.

soumith commented:
this looks largely good to go. It has merge conflicts, because I just merged in a PR that refactors the CMakeLists. Look at this PR: https://github.com/torch/cutorch/pull/174

szagoruyko commented:
I think all cmake MAGMA logic has to go to lib/THC and FindMAGMA has to be in lib/THC/cmake/CMakeLists.txt
btw should there be a doc how to install it?

dominikgrewe commented on lib/THC/THCTensorMathMagma.cu:
```diff
@@ -0,0 +1,472 @@
+
+  const int len = n*n;
+  cudaStream_t stream = THCState_getCurrentStream(state);
+  THCudaTensor_copyUpperSymmetric<<<(len+127/128), 128, 0, stream>>>(input_data, n, len);
```
Make sure you don't launch more than the maximum number of thread blocks. You can query it from `state->deviceProperties[currentDevice].maxGridSize[0]`.

dominikgrewe commented on lib/THC/THCTensorMathMagma.cu:
```diff
@@ -0,0 +1,472 @@
+  long stride[1] = { 1 };
+  THCudaTensor_rawResize(state, self, 1, size, stride);
+  size_t len = k * sizeof(float);
+  THCudaCheck(cudaMemcpy(self->storage->data + self->storageOffset, src, len, cudaMemcpyHostToDevice));
```
Make sure that `self` is actually contiguous. `THCudaTensor_rawResize` doesn't guarantee that.

Same in the function below.

colesbury replied:
I believe it does ensure that self is contiguous since it ensures the stride is 1.

dominikgrewe replied:
Fair point. Didn't think about the stride you're passing in.

colesbury commented:
I've fixed the bug @dominikgrewe pointed out where a large input could result in the block dim overflowing.

I also rebased and THC should still be separately buildable. The FindMAGMA is required by both cutorch and THC since THCGeneral.h.in defines USE_MAGMA based on whether the package is found.

soumith commented:
Are there any more reservations in merging this? Sam addressed dominik's and sergey's comments.

soumith commented:
@colesbury merge conflicts again :(

colesbury commented:
I rebased
